File: frontend/cli/src/app/persona/service/persona.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { DataPackage } from '../../models/data-package';
import { Persona } from '../../models/persona';
import { ReporteResponse } from '../../models/reporte';

/**
 * Servicio para gestión de personas y generación de reportes del sistema.
 * 
 * Proporciona operaciones CRUD completas para personas (docentes y personal),
 * además de funcionalidades especializadas para generación de reportes
 * individuales y de concepto general por año académico.
 * 
 * @author Mariano Bonansea
 * @version 1.0
 */
@Injectable({
  providedIn: 'root'
})
export class PersonaService {
  private personasUrl = "https://partes-docentes-1.onrender.com/personas";

  /**
   * Constructor del servicio.
   * 
   * @param http Cliente HTTP de Angular para realizar peticiones al backend
   */
  constructor(private http: HttpClient) { }

  /**
   * Obtiene todas las personas del sistema.
   * 
   * @returns Observable con el paquete de datos conteniendo la lista completa de personas
   */
  all(): Observable<DataPackage> {
    return this.http.get<DataPackage>(encodeURI(this.personasUrl));
  }

  /**
   * Obtiene una persona específica por su ID.
   * 
   * @param id Identificador único de la persona
   * @returns Observable con el paquete de datos conteniendo la persona solicitada
   */
  get(id: number): Observable<DataPackage> {
    return this.http.get<DataPackage>(encodeURI(`${this.personasUrl}/${id}`));
  }

  /**
   * Guarda una persona en el sistema (creación o actualización).
   * 
   * @param persona Objeto persona a guardar
   * @param isNew Indica si es una creación (true) o actualización (false)
   * @returns Observable con el resultado de la operación
   */
  save(persona: Persona, isNew: boolean = false): Observable<DataPackage> {
    return isNew
      ? this.http.post<DataPackage>(encodeURI(this.personasUrl), persona)
      : this.http.put<DataPackage>(encodeURI(this.personasUrl), persona);
  }

  /**
   * Elimina una persona del sistema.
   * 
   * @param id Identificador único de la persona a eliminar
   * @returns Observable con el resultado de la operación de eliminación
   */
  remove(id: number): Observable<DataPackage> {
    return this.http.delete<DataPackage>(encodeURI(`${this.personasUrl}/${id}`));
  }

  /**
   * Obtiene personas paginadas con ordenamiento personalizable.
   * 
   * @param page Número de página (base 1)
   * @param size Cantidad de elementos por página
   * @param sortField Campo por el cual ordenar (por defecto 'id')
   * @param sortDirection Dirección del ordenamiento ('asc' o 'desc')
   * @returns Observable con la página de personas solicitada
   */
  byPage(page: number, size: number, sortField: string = 'id', sortDirection: string = 'desc'): Observable<DataPackage> {
    return this.http.get<DataPackage>(
      encodeURI(`${this.personasUrl}/page?page=${page - 1}&size=${size}&sortField=${sortField}&sortDirection=${sortDirection}`)
    );
  }

  /**
   * Busca personas que coincidan con el término de búsqueda.
   * 
   * @param searchTerm Término de búsqueda para filtrar personas
   * @returns Observable con las personas que coinciden con la búsqueda
   */
  search(searchTerm: string): Observable<DataPackage> {
    return this.http.get<DataPackage>(encodeURI(`${this.personasUrl}/search/${searchTerm}`));
  }

  /**
   * Obtiene el reporte para un docente en un año específico
   * @param dni DNI del docente
   * @param anio Año del reporte
   * @returns Observable con la respuesta del reporte
   */
  obtenerReporte(dni: number, anio: number): Observable<ReporteResponse> {
    return this.http.get<ReporteResponse>(encodeURI(`${this.personasUrl}/dni/${dni}/reporte/${anio}`));
  }

  /**
   * Obtiene el reporte de concepto general para un año específico
   * @param anio Año del reporte
   * @returns Observable con la respuesta del reporte de concepto
   */
  obtenerReporteConcepto(anio: number): Observable<DataPackage> {
    return this.http.get<DataPackage>(encodeURI(`${this.personasUrl}/reporte-concepto/${anio}`));
  }
}

File: backend/src/main/java/unpsjb/labprog/backend/business/service/PersonaService.java
package unpsjb.labprog.backend.business.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import jakarta.transaction.Transactional;
import unpsjb.labprog.backend.business.repository.PersonaRepository;
import unpsjb.labprog.backend.model.Persona;

/**
 * Servicio para la gestión de personas del sistema. Proporciona operaciones
 * CRUD y búsqueda para la entidad Persona, incluyendo consultas por DNI, CUIL y
 * búsqueda por términos.
 *
 * @author Mariano Bonansea
 * @version 1.0
 */
@Service
public class PersonaService {

    @Autowired
    private PersonaRepository repository;

    /**
     * Obtiene todas las personas registradas en el sistema.
     *
     * @return lista de todas las personas
     */
    public List<Persona> findAll() {
        return repository.findAll();
    }

    /**
     * Busca una persona por su identificador único.
     *
     * @param id identificador de la persona
     * @return persona encontrada o null si no existe
     */
    public Persona findById(int id) {
        return repository.findById(id).orElse(null);
    }

    /**
     * Busca una persona por su número de DNI.
     *
     * @param dni número de DNI de la persona
     * @return persona encontrada o null si no existe
     */
    public Persona findByDni(long dni) {
        return repository.findByDni(dni).orElse(null);
    }

    /**
     * Busca una persona por su CUIL.
     *
     * @param cuil código único de identificación laboral
     * @return persona encontrada o null si no existe
     */
    public Persona findByCuil(String cuil) {
        return repository.findByCuil(cuil).orElse(null);
    }

    /**
     * Guarda una persona nueva o actualiza una existente.
     *
     * @param e persona a guardar
     * @return persona guardada
     */
    @Transactional
    public Persona save(Persona e) {
        return repository.save(e);
    }

    /**
     * Elimina una persona del sistema.
     *
     * @param id identificador de la persona a eliminar
     */
    @Transactional
    public void delete(int id) {
        repository.deleteById(id);
    }

    /**
     * Busca personas que coincidan con un término de búsqueda. La búsqueda es
     * insensible a mayúsculas y minúsculas.
     *
     * @param term término de búsqueda
     * @return lista de personas que coinciden con el término
     */
    public List<Persona> search(String term) {
        return repository.search("%" + term.toUpperCase() + "%");
    }

    /**
     * Obtiene una página de personas con paginación y ordenamiento
     * personalizado. Valida que el campo de ordenamiento sea permitido por
     * seguridad.
     *
     * @param page índice de página (basado en cero)
     * @param size tamaño de la página
     * @param sortField campo por el cual ordenar
     * @param sortDirection dirección del ordenamiento (asc o desc)
     * @return página de personas con el ordenamiento especificado
     */
    public Page<Persona> findByPage(int page, int size, String sortField, String sortDirection) {
        String[] allowedFields = {"id", "dni", "nombre", "apellido", "cuil", "titulo", "sexo", "domicilio", "telefono"};
        boolean isValidField = false;
        for (String field : allowedFields) {
            if (field.equals(sortField)) {
                isValidField = true;
                break;
            }
        }

        if (!isValidField) {
            sortField = "id";
        }

        Sort.Direction direction;
        if ("asc".equalsIgnoreCase(sortDirection)) {
            direction = Sort.Direction.ASC;
        } else {
            direction = Sort.Direction.DESC;
        }

        return repository.findAll(PageRequest.of(page, size, Sort.by(direction, sortField)));
    }
}
